### ISSUE TITLE
S1481 False Positive "Remove this useless assignment to local variable 'x'"

### ISSUE THREAD

**Issue by carldebilly** [SonarSource/sonar-dotnet#3169]
### Description

Please provide a succinct description of your issue. Also please provide the rule ID (e.g. S1234)

### Repro steps
Following code:
``` csharp
var x = double.Parse(strX);
var y = double.Parse(strY);
yield return (something, (x, y)); // Will trigger S1481 here on both x & y
```
As seen there: https://app.codacy.com/gh/unoplatform/uno/file/42594770330/issues/source?bid=16039680&fileBranchId=16659615#l63

### Expected behavior
Variables are correctly used there, false positive.

### Actual behavior
S1481 raised.

### Known workarounds
Ignore it.

### Related information

* SonarC# Version: `8.4` (the version used by _Codacy_)

**Comment by costin-zaharia-sonarsource** [comment]
Hi @carldebilly, 

thanks for the feedback.

I've tried to create a reproducer with the following:
`
public IEnumerable<(string, (double, double))> Tuples(string strX, string strY)
{
    var x = double.Parse(strX);
    var y = double.Parse(strY);
    yield return ("", (x, y)); // Will trigger S1481 here on both x & y
}
`
but this will not raise the warning and the link you provided is no longer valid.

Could you please help us with more context?

**Comment by Rubbiroid** [comment]
Have same issue in our project:
`public void StoreEvents(IList<IEvent> events)
{
	var debouncedEventTypes = new HashSet<Type>();
	var debouncedEvents = events.Where(NotYetDebounced);

	foreach (var debouncedEvent in debouncedEvents)
	{
		StoreEvent(debouncedEvent);
	}

	bool NotYetDebounced(IEvent @event)
	{
		return !(@event is IWithDebounce) || debouncedEventTypes.Add(@event.GetType());
	}
}`

sonar says that debouncedEventTypes  is not used: Remove this useless assignment to local variable S1854

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Rubbiroid, 

I can confirm your example as False Positive and I've created #3719 that you can track.

Your case is not related to original example with tuples in this issue.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @carldebilly,

Your issue is not reproducible using our analyzer in the way it's supposed to be used. Codacy uses file-by-file analysis with no project references information. Like that, each file is compiled separately as a standalone projects with many silently ignored compilation errors and almost no semantic model. That results in many false positives and even more valuable issues that are not detected.

You can consume our analyzer as a separate [Nuget Package](https://www.nuget.org/packages/SonarAnalyzer.CSharp/) in your project, with [SonarLint](https://www.sonarlint.org/) extension in your IDE or as a instance of [SonarQube](https://www.sonarqube.org/) or [SonarCloud](https://sonarcloud.io/) to experience full integration and pull request analysis with your build pipeline.
